drivers/misc/microbit_display/microbit_display.c

```diff
@@ -0,0 +1,514 @@
+/* SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) 2021 Peter Niebert, Aix-Marseille University
+ *
+ *
+ * New implementation of mb_display.h for compatibility with Microbit V2.
+ * This implementation uses a hardware timer of the Nordic SOC for periodic
+ * interrupts independent of Kernel timing for image refresh.
+ * The animation is timed separately by delayed work submitted to the kernel.
+ *
+ * References:
+ *
+ * https://www.microbit.co.uk/device/screen
+ * https://lancaster-university.github.io/microbit-docs/ubit/display/
+ * https://github.com/zephyrproject-rtos/zephyr/blob/main/include/...
+ * ...display/mb_display.h
+ */
+
+#include <sys/__assert.h>
+
+#include <zephyr.h>
+#include <init.h>
+#include <logging/log.h>
+#include <sw_isr_table.h>
+
+#include <stdint.h>
+#include <stdarg.h>
+
+#include <microbit_display.h>
+#include <mb_font.h>
+
+#include <nrfx_timer.h>
+#include <nrfx_ppi.h>
+#include <hal/nrf_gpio.h>
+#include <nrfx_gpiote.h>
+
+LOG_MODULE_REGISTER(mb2_display, LOG_LEVEL_DBG);
+
+/************ pin related definitions for micro:bit V1 and V2 *********/
+
+#define DISPLAY_ROWS 5
+#define DISPLAY_COLS 5
+
+#ifdef CONFIG_BOARD_BBC_MICROBIT
+
+static const uint16_t led_column_row[25] = { 13 << 8 | 4, 14 << 8 | 7,
+	13 << 8 | 5, 14 << 8 | 8, 13 << 8 | 6, 15 << 8 | 7,	 15 << 8 | 8,
+	15 << 8 | 9, 15 << 8 | 10, 15 << 8 | 11, 14 << 8 | 5, 13 << 8 | 12,
+	14 << 8 | 6, 15 << 8 | 12, 14 << 8 | 4, 13 << 8 | 11,
+	13 << 8 | 10, 13 << 8 | 9,	 13 << 8 | 8, 13 << 8 | 7,
+	15 << 8 | 6, 14 << 8 | 10, 15 << 8 | 4, 14 << 8 | 9,
+	15 << 8 | 5 };
+
+static const uint32_t row_mask0 =
+	1 << 4 | 1 << 5 | 1 << 6 | 1 << 7 | 1 << 8 | 1 << 9 | 1 << 10 | 1 << 11 | 1 << 12;
+
+static const uint32_t col_mask = 1 << 13 | 1 << 14 | 1 << 15;
+
+/* use low level inline function for pin set */
+#define mypinset nrf_gpio_pin_set
+#define mypinclear nrf_gpio_pin_clear
+
+#else
+#ifdef CONFIG_BOARD_BBC_MICROBIT_V2
+
+/* columns/anodes followed by rows/cathodes */
+
+static const uint16_t led_column_row[25] = {
+	21 << 8 | 28, 21 << 8 | 11, 21 << 8 | 31, 21 << 8 | 37, 21 << 8 | 30,
+	22 << 8 | 28, 22 << 8 | 11, 22 << 8 | 31, 22 << 8 | 37, 22 << 8 | 30,
+	15 << 8 | 28, 15 << 8 | 11, 15 << 8 | 31, 15 << 8 | 37, 15 << 8 | 30,
+	24 << 8 | 28, 24 << 8 | 11, 24 << 8 | 31, 24 << 8 | 37, 24 << 8 | 30,
+	19 << 8 | 28, 19 << 8 | 11, 19 << 8 | 31, 19 << 8 | 37, 19 << 8 | 30
```

Comment:
Are these pin numbers? If so, would it be possible to define them in device tree, instead of source files?